gossip/src/duplicate_shred_handler.rs

```diff
@@ -63,10 +60,17 @@ pub struct DuplicateShredHandler {
     // We don't want bad guys to inflate the chunk map, so we limit the number of
     // pending proofs from each pubkey to ALLOWED_SLOTS_PER_PUBKEY.
     validator_pending_proof_map: HashMap<Pubkey, HashSet<Slot>>,
-    // remember the last root slot handled, clear anything older than last_root.
-    last_root: Slot,
+    // Cache last root to reduce read lock.
+    last_root: Option<Slot>,
+    // remember the last root slot cleaned, clear anything between last_root and last_root_cleaned.
+    last_root_cleaned: Option<Slot>,
     blockstore: Arc<Blockstore>,
     leader_schedule_cache: Arc<LeaderScheduleCache>,
+    bank_forks: Arc<RwLock<BankForks>>,
+    // Cache information from root bank so we could function correctly without reading roots.
+    cached_on_slot: Slot,
+    cached_staked_nodes: Option<Arc<HashMap<Pubkey, u64>>>,
```

Comment:
Done

Reply:
`cached_staked_nodes` does not need to be an `Option`.
If epoch-stakes are none, we already have bigger problems.
Just use `unwrap_or_default`. Maybe error log if it is none.

Reply:
Before it is initialized the first time I need to give it a value, Option seems most natural here, using None to indicate not initialized so I should read epoch info before use.

Reply:
This handler is initialized in TVU initialization, I'm not sure there is a valid bank to read from at that point, also it's better not to obtain read locks in initialization logic any way.

Option seems most natural, it also makes copying from root_bank.epoch_staked_nodes more straightforward. But I can live with an empty stake map at initialization as well, changed.

Reply:
> I'm not sure there is a valid bank to read from at that point, 

If there is not, then we need to understand when that would be the case and why.
My understanding from the `BankForks` api is that it always have a root-bank; and the rest here are obtained from the root_bank.

> also it's better not to obtain read locks in initialization logic any way.

why not?

Reply:
@wen-coding I am still curious to know what the problem is to fully populating the struct fields at initialization.

Reply:
The main concern is separation of concerns. I guess I'm extremely against long read/write access in initialization because I'd been bitten by this before in emergency, we needed some quick fix to go into production to rescue the world, but the initialization logic spent x minutes reading the disk before accepting requests, so that's x minutes of requests lost.

All initialization logic are simple at first, but as time goes by they will get more and more bloated, until one day you need to spend a lot of time deciding "A needs to be constructed before B, and B needs to be constructed before C". If you put real dependencies in initialize(), then you can construct the objects in any order you want, then during initialization phase you can initialize according to the real dependencies. It might not seem to be a lot of difference, but it makes life a bit easier because very often you can't even construct a complex object without all its dependency constructed.

It also helps testing, because if you put all the physical world dependencies in a separate initialize() logic, then you can easily stick in mock objects after construction without calling initialize(), your tests will not have to jump through the hoops constructing the fake world so the constructor can give you an object.

That said, it is hard to draw the line somewhere, but generally I prefer lazy initialization and make the constructor small.

Reply:
If there are particular cases that initialization causes issues we can address those specifically.
But I think by default better initialize types correctly rather than leave them in a limbo state.
Specially all `Option<..>` which can be none will make the code convoluted and easily may introduce bugs.